lib/commands/recordscreen.js

```diff
@@ -1,61 +1,157 @@
-const SIMULATOR_PGREP_PATTERN = (udid) => `simctl io ${udid} recordVideo`;
-const DEFAULT_EXT = '.mp4';
-
-async function extractCurrentRecordingPath (pid) {
-  const {output} = await exec('ps', ['o', 'command', '-p', pid]);
-  log.debug(`Got the following output from ps: ${output}`);
-  const pattern = new RegExp(/[\s="'](\/.*\.mp4)/);
-  const matches = pattern.exec(output);
-  return _.isEmpty(matches) ? null : _.last(matches);
-}
+const DEFAULT_MJPEG_SERVER_PORT = 9100;
+const MP4_EXT = '.mp4';
+const DEFAULT_FPS = 10;
+const FFMPEG_BINARY = 'ffmpeg';
 
-async function finishScreenCapture (pid) {
-  try {
-    await exec('kill', ['-2', pid]);
-  } catch (e) {
-    return true;
+
+class ScreenRecorder {
+  constructor (udid, videoPath, opts = {}) {
+    this.videoPath = videoPath;
+    this.opts = opts;
+    this.udid = udid;
+    this.mainProcess = null;
+    this.iproxy = null;
+    this.timeoutHandler = null;
   }
-  try {
-    await waitForCondition(async () => {
+
+  async start (timeout) {
+    try {
+      await fs.which(FFMPEG_BINARY);
+    } catch (err) {
+      throw new Error(`'${FFMPEG_BINARY}' binary is not found in PATH. Install it using 'brew install ffmpeg'. ` +
+        `Check https://www.ffmpeg.org/download.html for more details.`);
+    }
+
+    const localPort = this.opts.remotePort;
+    if (this.opts.usePortForwarding) {
+      await this.startIproxy(localPort);
+    }
+
+    const args = [
+      '-f', 'mjpeg',
+      '-r', this.opts.videoFps,
+      '-i', `http://localhost:${localPort}`,
+      '-vcodec', 'mjpeg',
+      '-y', this.videoPath,
+    ];
+    this.mainProcess = new SubProcess(FFMPEG_BINARY, args);
+    this.mainProcess.on('output', (stdout, stderr) => {
+      if (stderr && !stderr.includes('frame=')) {
+        log.warn(`[${FFMPEG_BINARY}] ${stderr}`);
+      }
+    });
+    // Give ffmpeg some time for init
+    await this.mainProcess.start(5000);
+    log.info(`Starting screen capture on the device '${this.udid}' with command: '${FFMPEG_BINARY} ${args.join(' ')}'. ` +
+      `Will timeout in ${timeout}ms`);
+
+    this.timeoutHandler = setTimeout(async () => {
+      if (!await this.interrupt()) {
+        log.warn(`Cannot finish the active screen recording on the device '${this.udid}' after ${timeout}ms timeout`);
+      }
+    }, timeout);
+  }
+
+  async startIproxy (localPort) {
+    this.iproxy = new iProxy(this.udid, localPort, this.opts.remotePort);
+    try {
+      await this.iproxy.start();
+    } catch (err) {
+      log.warn(`Cannot start iproxy. Assuming it is already forwarding the remote port ${this.opts.remotePort} to ${localPort} ` +
+        `for the device ${this.udid}. Set the custom value to 'mjpegServerPort' capability if this is an undesired behavior. ` +
+        `Original error: ${err.message}`);
+      this.iproxy = null;
+    }
+  }
+
+  async stopIproxy () {
+    if (!this.iproxy) {
+      return;
+    }
+
+    const quitPromise = this.iproxy.quit();
+    this.iproxy = null;
```

Comment:
this could just be moved into a `finally` block right?

Reply:
this was deliberate change to avoid race condition with setTimeout cal (`await` switches the flow to pick up the next task from the run loop)

Reply:
I refactored it anyway